assets/js/payment-methods-demo/express-payment/apple-pay/apple-pay-express.js

```diff
@@ -0,0 +1,380 @@
+	getPaymentMethodData,
+	normalizeShippingAddressForCheckout,
+	normalizeShippingOptions,
+	normalizeLineItems,
+} from './normalize';
+import {
+	getStripeServerData,
+	getPaymentRequest,
+	updatePaymentRequest,
+	canDoApplePay,
+	getTotalPaymentItem,
+	stripePromise,
+} from '../../stripe-utils';
+
+/**
+ * External dependencies
+ */
+import { useRef, useState, useEffect } from '@wordpress/element';
+import {
+	Elements,
+	PaymentRequestButtonElement,
+	useStripe,
+} from '@stripe/react-stripe-js';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * @typedef {import('../../stripe-utils/type-defs').Stripe} Stripe
+ * @typedef {import('../../stripe-utils/type-defs').StripePaymentRequest} StripePaymentRequest
+ * @typedef {import('@woocommerce/type-defs/registered-payment-method-props').RegisteredPaymentMethodProps} RegisteredPaymentMethodProps
+ */
+
+/**
+ * @typedef {Object} WithStripe
+ *
+ * @property {Stripe} [stripe] Stripe api (might not be present)
+ */
+
+/**
+ * @typedef {RegisteredPaymentMethodProps & WithStripe} StripeRegisteredPaymentMethodProps
+ */
+
+// @todo, note it should be trivial to implement cross browser payment api
+// handling here:
+// - ApplePay in Safari
+// - ChromePay in Chrome
+// - not supported for other browsers yet (but if assuming stripe implements the
+//   official PaymentRequest API in their library this should enable support!).
+
+/**
+ * AppleExpressComponent
+ *
+ * @param {StripeRegisteredPaymentMethodProps} props Incoming props
+ */
+const ApplePayExpressComponent = ( {
+	paymentStatus,
+	shippingData,
+	billing,
+	eventRegistration,
+	onSubmit,
+	activePaymentMethod,
+	setActivePaymentMethod,
+} ) => {
+	/**
+	 * @type {[ StripePaymentRequest|null, function( StripePaymentRequest ):StripePaymentRequest|null]}
+	 */
+	// @ts-ignore
+	const [ paymentRequest, setPaymentRequest ] = useState( null );
+	const stripe = useStripe();
+	const [ canMakePayment, setCanMakePayment ] = useState( false );
+	const eventHandlers = useRef( DEFAULT_STRIPE_EVENT_HANDLERS );
+	const currentBilling = useRef( billing );
+	const currentShipping = useRef( shippingData );
+	const currentPaymentRequest = useRef( paymentRequest );
+	const currentPaymentStatus = useRef( paymentStatus );
+	const currentEventRegistration = useRef( eventRegistration );
+	const isActive = useRef( activePaymentMethod === PAYMENT_METHOD_NAME );
+
+	// update refs when any change
+	useEffect( () => {
+		currentBilling.current = billing;
+		currentShipping.current = shippingData;
+		currentPaymentRequest.current = paymentRequest;
+		currentPaymentStatus.current = paymentStatus;
+		currentEventRegistration.current = eventRegistration;
+		isActive.current = activePaymentMethod === PAYMENT_METHOD_NAME;
+	}, [ billing, shippingData, paymentRequest, paymentStatus, isActive ] );
+
+	//set paymentRequest
+	useEffect( () => {
+		// can't do anything if stripe isn't available yet or we have zero total
+		if ( ! stripe || ! billing.cartTotal.value ) {
```

Comment:
We should probably expose order and payment info on the CheckoutProvider rather than cart at some point.

Reply:
Agreed, but I also want to try to keep the payment interface usable for the cart context too (even though we won't implement that yet). Reason being is that _express payment methods_ should be usable for cart or "pay now" type buttons in the shop.